npm/generate_katas_content.js

```diff
@@ -60,174 +49,250 @@ function getTitleFromMarkdown(markdown) {
-    referenceImplementation: referenceSource,
-    verificationImplementation: verificationSource,
-    contentAsMarkdown: contentAsMarkdown,
-    contentAsHtml: contentAsHtml,
+    id: properties.id,
+    codeDependencies: codeDependencies,
+    verificationCode: verificationCode,
+    placeholderCode: placeholderCode,
     solutionAsMarkdown: solutionAsMarkdown,
     solutionAsHtml: solutionAsHtml,
   };
 }
 
-function buildReadingContent(id, path) {
-  const contentAsMarkdown = readFileSync(
-    join(path, contentFileNames.index),
-    "utf8"
+function generateMacroSection(kataPath, match, globalCodeSources) {
+  const type = match.groups.type;
+  const propertiesJson = match.groups.json;
+  const properties = tryParseJSON(
+    propertiesJson,
+    `Invalid JSON for ${type} macro.`
   );
-  const contentAsHtml = marked.parse(contentAsMarkdown);
-  const title = getTitleFromMarkdown(contentAsMarkdown);
-  return {
-    type: "reading",
-    id: id,
-    title: title,
-    contentAsMarkdown: contentAsMarkdown,
-    contentAsHtml: contentAsHtml,
-  };
-}
-
-function symmetricDifference(setA, setB) {
-  const difference = new Set(setA);
-  for (const elem of setB) {
-    if (difference.has(elem)) {
-      difference.delete(elem);
-    } else {
-      difference.add(elem);
-    }
+  if (type === "example") {
+    return generateExampleSection(kataPath, properties);
+  } else if (type === "exercise") {
+    return generateExerciseSection(kataPath, properties, globalCodeSources);
   }
-  return difference;
+
+  throw new Error(`Unknown macro type ${type}`);
 }
 
-function getItemType(path) {
-  const itemTypeFileSets = {
-    reading: new Set([contentFileNames.index]),
-    example: new Set([contentFileNames.index, contentFileNames.qsharpExample]),
-    exercise: new Set([
-      contentFileNames.index,
-      contentFileNames.qsharpPlaceholder,
-      contentFileNames.qsharpSolution,
-      contentFileNames.qsharpVerify,
-      contentFileNames.textSolution,
-    ]),
+function generateTextSection(markdown) {
+  const html = marked.parse(markdown);
+  return {
+    type: "text",
+    contentAsMarkdown: markdown,
+    contentAsHtml: html,
   };
-
-  const itemFiles = new Set(readdirSync(path));
-  return Object.keys(itemTypeFileSets).find(function (key) {
-    const fileSet = itemTypeFileSets[key];
-    return symmetricDifference(fileSet, itemFiles).size === 0;
-  });
 }
 
-function buildItemContent(path) {
-  const itemId = `${basename(dirname(path))}__${basename(path)}`;
-  const itemType = getItemType(path);
-  if (itemType === "example") {
-    return buildExampleContent(itemId, path);
-  } else if (itemType === "exercise") {
-    return buildExerciseContent(itemId, path);
-  } else if (itemType === "reading") {
-    return buildReadingContent(itemId, path);
+function generateSections(kataPath, markdown, globalCodeSources) {
+  const sections = [];
+  const macroRegex = /@\[(?<type>\w+)\]\(\s*(?<json>\{[^@]+\})\s*\)\s+/g;
```

Comment:
This regular expression needs some work. It's using a greedy match, so for example if you have the below text in the file

```
@[exercise]({
    "id": "preparing_bell_state",
    "other stuff": true
})

({random content})
```

It will match all of it including the random content. (As long as it has a trailing newline - I'm not sure if the final `\s+` needs to be there either)

Reply:
I believe something like the below should work well. It's non-greedy, and terminates the JSON match when it sees the chars `})` followed by a new-line (which seems like a reasonable termination pattern that won't happen inside the JSON). I removed some of the allowed white-space (which seemed unnecesary) and added the dot flag also (`/s`) so that dot matches newlines.

`/@\[(?<type>\w+)\]\((?<json>\{.*?\})\)\n/gs`